docs/mypy_plugin.md

```diff
@@ -0,0 +1,137 @@
+if your field names or types change.
+
+There are other benefits too! See below for more details.
+
+### Plugin Capabilities
+
+#### Generate a signature for `Model.__init__`
+* Any required fields that don't have dynamically-determined aliases will be included as required
+  keyword arguments.
+* If `Config.allow_population_by_field_name=True`, the generated signature will use the field names,
+  rather than aliases.
+* For subclasses of `BaseSettings`, all fields are treated as optional since they may be read from the environment.
+* If `Config.extra="forbid"` and you don't make use of dynamically-determined aliases, the generated signature
+  will not allow unexpected inputs.
+* **Optional:** If the `init_allow_extra` **plugin setting** (see below) is set to `False`, unexpected inputs to
+  `__init__` will raise errors even if `Config.extra` is not `"forbid"`.
+* **Optional:** If the `init_typed` **plugin setting** (see below) is set to `True`, the generated signature
+  will use the types of the model fields (otherwise they will be annotated as `Any` to allow parsing).
+ 
+#### Generate a typed signature for `Model.construct`
+* This method is a (potentially much faster) alternative to `__init__` when input data is known to be valid,
+  and does not need to be parsed.
+* However, this method performs no runtime validation, so static checks are important to detect errors.
+
+#### Respect `Config.allow_mutation`
+* If `Config.allow_mutation` is `False`, you'll get a mypy error if you try to change the value of a model field.
+
+#### Respect `Config.orm_mode`
+* If `Config.orm_mode` is `False`, you'll get a mypy error if you try to call `.from_orm()`.
+ 
+### Optional Capabilites:
+#### Prevent the use of required dynamic aliases
+* If the `allow_required_dynamic_aliases` **plugin setting** (see below) is set to `False`, you'll get a mypy
+  error any time you use a dynamically-determined alias or alias generator on a model with
+`Config.allow_population_by_field_name=False`.
+* This is important because if such aliases are present, mypy cannot properly type check calls to `__init__`.
+  In this case, it will default to treating all arguments as optional. 
+
+#### Prevent the use of untyped fields
+* If the `allow_untyped_fields` **plugin setting** (see below) is set to `False`, you'll get a mypy error
+  any time you create a field on a model without annotating its type.
+* This is important because non-annotated fields may result in **validators being applied in a surprising order**.
+* In addition, mypy may not be able to correctly infer the type of the field, and may miss
+  checks or raise spurious errors.
+
+### Enabling the Plugin
+
+To enable the plugin, just add `pydantic.mypy` to the list of plugins in your mypy config file.
+
+For example, a `mypy.ini` file with the pydantic mypy plugin enabled might look like this:
+```ini
+[mypy]
+plugins = pydantic.mypy
```

Comment:
I wanted to introduce the complexity one bit at a time, and you don't need to worry about the `[pydantic-mypy]` section unless you want to configure special options.

I see two options: 1) I could add a link to the section below with the `[pydantic-mypy]` section, or 2) I could add the section here and set the values to the defaults.

Reply:
make sense.

Maybe:
* remove the `follow_imports...disallow_untyped_defs = True` part since that confuses this simple example
* add a link to `/usage/mypy/` saying something like "go here for more details on mypy usage"